File: src-tauri/src/db/invoice.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

use super::Database;

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Invoice {
    pub id: String,
    #[serde(rename = "organizationId")]
    #[sqlx(rename = "organizationId")]
    pub organization_id: String,
    pub number: String,
    pub state: String,
    #[serde(rename = "clientId")]
    #[sqlx(rename = "clientId")]
    pub client_id: String,
    pub date: i64,
    #[serde(rename = "dueDate")]
    #[sqlx(rename = "dueDate")]
    pub due_date: Option<i64>,
    pub currency: String,
    #[serde(rename = "customerNotes")]
    #[sqlx(rename = "customerNotes")]
    pub customer_notes: Option<String>,
    pub total: i64,  // Stored as cents
    #[serde(rename = "taxTotal")]
    #[sqlx(rename = "taxTotal")]
    pub tax_total: i64,  // Stored as cents
    #[serde(rename = "subTotal")]
    #[sqlx(rename = "subTotal")]
    pub sub_total: i64,  // Stored as cents
    #[serde(rename = "createdAt")]
    #[sqlx(rename = "createdAt")]
    pub created_at: Option<String>,
    // Only for joined queries
    #[serde(rename = "clientName")]
    #[sqlx(rename = "clientName")]
    pub client_name: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct InvoiceLineItem {
    pub id: String,
    #[serde(rename = "invoiceId")]
    #[sqlx(rename = "invoiceId")]
    pub invoice_id: String,
    pub description: Option<String>,
    pub quantity: f64,
    #[serde(rename = "unitPrice")]
    #[sqlx(rename = "unitPrice")]
    pub unit_price: i64,  // Stored as cents
    #[serde(rename = "taxRate")]
    #[sqlx(rename = "taxRate")]
    pub tax_rate: Option<String>,
    #[serde(rename = "createdAt")]
    #[sqlx(rename = "createdAt")]
    pub created_at: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CreateInvoiceRequest {
    pub id: String,
    #[serde(rename = "organizationId")]
    pub organization_id: String,
    pub number: String,
    pub state: String,
    #[serde(rename = "clientId")]
    pub client_id: String,
    pub date: i64,
    #[serde(rename = "dueDate")]
    pub due_date: Option<i64>,
    pub currency: String,
    #[serde(rename = "customerNotes")]
    pub customer_notes: Option<String>,
    pub total: i64,  // Stored as cents
    #[serde(rename = "taxTotal")]
    pub tax_total: i64,  // Stored as cents
    #[serde(rename = "subTotal")]
    pub sub_total: i64,  // Stored as cents
    #[serde(rename = "lineItems")]
    pub line_items: Vec<CreateInvoiceLineItemRequest>,
}

#[derive(Debug, Deserialize)]
pub struct CreateInvoiceLineItemRequest {
    pub description: Option<String>,
    pub quantity: f64,
    #[serde(rename = "unitPrice")]
    pub unit_price: f64,
    #[serde(rename = "taxRate")]
    pub tax_rate: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateInvoiceRequest {
    pub number: Option<String>,
    pub state: Option<String>,
    #[serde(rename = "clientId")]
    pub client_id: Option<String>,
    pub date: Option<i64>,
    #[serde(rename = "dueDate")]
    pub due_date: Option<i64>,
    pub currency: Option<String>,
    #[serde(rename = "customerNotes")]
    pub customer_notes: Option<String>,
    pub total: Option<i64>,  // Stored as cents
    #[serde(rename = "taxTotal")]
    pub tax_total: Option<i64>,  // Stored as cents
    #[serde(rename = "subTotal")]
    pub sub_total: Option<i64>,  // Stored as cents
    #[serde(rename = "lineItems")]
    pub line_items: Option<Vec<CreateInvoiceLineItemRequest>>,
}

impl Database {
    pub async fn get_invoices(&self, organization_id: &str) -> Result<Vec<Invoice>, sqlx::Error> {
        sqlx::query_as::<_, Invoice>(
            r#"
            SELECT
                invoices.*,
                clients.name AS clientName
            FROM
                invoices
            INNER JOIN
                clients ON invoices.clientId = clients.id
            WHERE
                invoices.organizationId = ?
            ORDER BY
                invoices.date DESC
            "#,
        )
        .bind(organization_id)
        .fetch_all(&self.pool)
        .await
    }

    pub async fn get_invoice(&self, invoice_id: &str) -> Result<Option<Invoice>, sqlx::Error> {
        sqlx::query_as::<_, Invoice>(
            r#"
            SELECT
                invoices.*,
                clients.name AS clientName
            FROM
                invoices
            INNER JOIN
                clients ON invoices.clientId = clients.id
            WHERE
                invoices.id = ?
            LIMIT 1
            "#,
        )
        .bind(invoice_id)
        .fetch_optional(&self.pool)
        .await
    }

    pub async fn get_invoice_line_items(&self, invoice_id: &str) -> Result<Vec<InvoiceLineItem>, sqlx::Error> {
        sqlx::query_as::<_, InvoiceLineItem>(
            r#"
            SELECT *
            FROM invoiceLineItems
            WHERE invoiceId = ?
            ORDER BY createdAt ASC
            "#,
        )
        .bind(invoice_id)
        .fetch_all(&self.pool)
        .await
    }

    pub async fn create_invoice(&self, invoice: CreateInvoiceRequest) -> Result<Invoice, sqlx::Error> {
        let mut tx = self.pool.begin().await?;

        // Insert invoice
        sqlx::query(
            r#"
            INSERT INTO invoices (
                id, organizationId, number, state, clientId, date, dueDate, 
                currency, customerNotes, total, taxTotal, subTotal
            )
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            "#,
        )
        .bind(&invoice.id)
        .bind(&invoice.organization_id)
        .bind(&invoice.number)
        .bind(&invoice.state)
        .bind(&invoice.client_id)
        .bind(&invoice.date)
        .bind(&invoice.due_date)
        .bind(&invoice.currency)
        .bind(&invoice.customer_notes)
        .bind(&invoice.total)
        .bind(&invoice.tax_total)
        .bind(&invoice.sub_total)
        .execute(&mut *tx)
        .await?;

        // Insert line items
        for line_item in &invoice.line_items {
            let line_item_id = nanoid::nanoid!();
            sqlx::query(
                r#"
                INSERT INTO invoiceLineItems (id, invoiceId, description, quantity, unitPrice, taxRate)
                VALUES (?, ?, ?, ?, ?, ?)
                "#,
            )
            .bind(&line_item_id)
            .bind(&invoice.id)
            .bind(&line_item.description)
            .bind(&line_item.quantity)
            .bind(&line_item.unit_price)
            .bind(&line_item.tax_rate)
            .execute(&mut *tx)
            .await?;
        }

        // Increment the invoice counter for the organization
        sqlx::query("UPDATE organizations SET invoice_number_counter = invoice_number_counter + 1 WHERE id = ?")
            .bind(&invoice.organization_id)
            .execute(&mut *tx)
            .await?;

        tx.commit().await?;

        self.get_invoice(&invoice.id).await?
            .ok_or_else(|| sqlx::Error::RowNotFound)
    }

    pub async fn update_invoice(
        &self,
        invoice_id: &str,
        updates: UpdateInvoiceRequest,
    ) -> Result<Invoice, sqlx::Error> {
        let mut tx = self.pool.begin().await?;

        // Update invoice fields (using all fields with Option checks)
        sqlx::query(
            r#"
            UPDATE invoices
            SET number = COALESCE(?, number),
                state = COALESCE(?, state),
                clientId = COALESCE(?, clientId),
                date = COALESCE(?, date),
                dueDate = COALESCE(?, dueDate),
                currency = COALESCE(?, currency),
                customerNotes = COALESCE(?, customerNotes),
                total = COALESCE(?, total),
                taxTotal = COALESCE(?, taxTotal),
                subTotal = COALESCE(?, subTotal)
            WHERE id = ?
            "#,
        )
        .bind(&updates.number)
        .bind(&updates.state)
        .bind(&updates.client_id)
        .bind(&updates.date)
        .bind(&updates.due_date)
        .bind(&updates.currency)
        .bind(&updates.customer_notes)
        .bind(&updates.total)
        .bind(&updates.tax_total)
        .bind(&updates.sub_total)
        .bind(invoice_id)
        .execute(&mut *tx)
        .await?;

        // Update line items if provided
        if let Some(line_items) = updates.line_items {
            // Delete existing line items
            sqlx::query("DELETE FROM invoiceLineItems WHERE invoiceId = ?")
                .bind(invoice_id)
                .execute(&mut *tx)
                .await?;

            // Insert new line items
            for line_item in &line_items {
                let line_item_id = nanoid::nanoid!();
                sqlx::query(
                    r#"
                    INSERT INTO invoiceLineItems (id, invoiceId, description, quantity, unitPrice, taxRate)
                    VALUES (?, ?, ?, ?, ?, ?)
                    "#,
                )
                .bind(&line_item_id)
                .bind(invoice_id)
                .bind(&line_item.description)
                .bind(&line_item.quantity)
                .bind(&line_item.unit_price)
                .bind(&line_item.tax_rate)
                .execute(&mut *tx)
                .await?;
            }
        }

        tx.commit().await?;

        self.get_invoice(invoice_id).await?
            .ok_or_else(|| sqlx::Error::RowNotFound)
    }

    pub async fn delete_invoice(&self, invoice_id: &str) -> Result<bool, sqlx::Error> {
        let mut tx = self.pool.begin().await?;

        // Delete line items first (foreign key constraint)
        sqlx::query("DELETE FROM invoiceLineItems WHERE invoiceId = ?")
            .bind(invoice_id)
            .execute(&mut *tx)
            .await?;

        // Delete invoice
        let result = sqlx::query("DELETE FROM invoices WHERE id = ?")
            .bind(invoice_id)
            .execute(&mut *tx)
            .await?;

        tx.commit().await?;

        Ok(result.rows_affected() > 0)
    }
}

File: db/invoice.go
package db

import (
	"fmt"

	gonanoid "github.com/matoous/go-nanoid/v2"
)

// Invoice mirrors the invoices table (with an optional joined clientName).
type Invoice struct {
	ID             string   `db:"id"             json:"id"`
	OrganizationID string   `db:"organizationId" json:"organizationId"`
	Number         string   `db:"number"         json:"number"`
	State          string   `db:"state"          json:"state"`
	ClientID       string   `db:"clientId"       json:"clientId"`
	Date           int64    `db:"date"           json:"date"`
	DueDate        *int64   `db:"dueDate"        json:"dueDate"`
	Currency       string   `db:"currency"       json:"currency"`
	CustomerNotes  *string  `db:"customerNotes"  json:"customerNotes"`
	OverdueCharge  *float64 `db:"overdueCharge"  json:"overdueCharge"`
	Total          int64    `db:"total"          json:"total"`
	TaxTotal       int64    `db:"taxTotal"       json:"taxTotal"`
	SubTotal       int64    `db:"subTotal"       json:"subTotal"`
	CreatedAt      *string  `db:"createdAt"      json:"createdAt"`
	ClientName     *string  `db:"clientName"     json:"clientName"`
}

// InvoiceLineItem mirrors the invoiceLineItems table.
type InvoiceLineItem struct {
	ID          string   `db:"id"          json:"id"`
	InvoiceID   string   `db:"invoiceId"   json:"invoiceId"`
	Description *string  `db:"description" json:"description"`
	Quantity    float64  `db:"quantity"    json:"quantity"`
	UnitPrice   int64    `db:"unitPrice"   json:"unitPrice"`
	TaxRate     *string  `db:"taxRate"     json:"taxRate"`
	Position    int      `db:"position"    json:"position"`
	CreatedAt   *string  `db:"createdAt"   json:"createdAt"`
}

// CreateInvoiceLineItemRequest is a single line item within a create/update request.
type CreateInvoiceLineItemRequest struct {
	Description *string  `json:"description"`
	Quantity    float64  `json:"quantity"`
	UnitPrice   float64  `json:"unitPrice"`
	TaxRate     *string  `json:"taxRate"`
}

// CreateInvoiceRequest is the payload for creating an invoice.
type CreateInvoiceRequest struct {
	ID             string                         `json:"id"`
	OrganizationID string                         `json:"organizationId"`
	Number         string                         `json:"number"`
	State          string                         `json:"state"`
	ClientID       string                         `json:"clientId"`
	Date           int64                          `json:"date"`
	DueDate        *int64                         `json:"dueDate"`
	Currency       string                         `json:"currency"`
	CustomerNotes  *string                        `json:"customerNotes"`
	OverdueCharge  *float64                       `json:"overdueCharge"`
	Total          int64                          `json:"total"`
	TaxTotal       int64                          `json:"taxTotal"`
	SubTotal       int64                          `json:"subTotal"`
	LineItems      []CreateInvoiceLineItemRequest `json:"lineItems"`
}

// UpdateInvoiceRequest is the payload for updating an invoice.
type UpdateInvoiceRequest struct {
	Number        *string                         `json:"number"`
	State         *string                         `json:"state"`
	ClientID      *string                         `json:"clientId"`
	Date          *int64                          `json:"date"`
	DueDate       *int64                          `json:"dueDate"`
	Currency      *string                         `json:"currency"`
	CustomerNotes *string                         `json:"customerNotes"`
	OverdueCharge *float64                        `json:"overdueCharge"`
	Total         *int64                          `json:"total"`
	TaxTotal      *int64                          `json:"taxTotal"`
	SubTotal      *int64                          `json:"subTotal"`
	LineItems     *[]CreateInvoiceLineItemRequest `json:"lineItems"`
}

func (d *Database) GetInvoices(organizationID string) ([]Invoice, error) {
	invoices := []Invoice{}
	err := d.DB.Select(&invoices, `
		SELECT invoices.*, clients.name AS clientName
		FROM invoices
		INNER JOIN clients ON invoices.clientId = clients.id
		WHERE invoices.organizationId = ?
		ORDER BY invoices.date DESC`,
		organizationID,
	)
	if err != nil {
		return nil, fmt.Errorf("get_invoices: %w", err)
	}
	return invoices, nil
}

func (d *Database) GetInvoice(invoiceID string) (*Invoice, error) {
	var invoice Invoice
	err := d.DB.Get(&invoice, `
		SELECT invoices.*, clients.name AS clientName
		FROM invoices
		INNER JOIN clients ON invoices.clientId = clients.id
		WHERE invoices.id = ?
		LIMIT 1`,
		invoiceID,
	)
	if err != nil {
		return nil, fmt.Errorf("get_invoice: %w", err)
	}
	return &invoice, nil
}

func (d *Database) GetInvoiceLineItems(invoiceID string) ([]InvoiceLineItem, error) {
	items := []InvoiceLineItem{}
	err := d.DB.Select(&items,
		`SELECT * FROM invoiceLineItems WHERE invoiceId = ? ORDER BY position ASC, createdAt ASC`,
		invoiceID,
	)
	if err != nil {
		return nil, fmt.Errorf("get_invoice_line_items: %w", err)
	}
	return items, nil
}

func (d *Database) CreateInvoice(req CreateInvoiceRequest) (*Invoice, error) {
	tx, err := d.DB.Beginx()
	if err != nil {
		return nil, fmt.Errorf("create_invoice begin: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	_, err = tx.Exec(`
		INSERT INTO invoices (
			id, organizationId, number, state, clientId, date, dueDate,
			currency, customerNotes, overdueCharge, total, taxTotal, subTotal
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		req.ID, req.OrganizationID, req.Number, req.State, req.ClientID,
		req.Date, req.DueDate, req.Currency, req.CustomerNotes, req.OverdueCharge,
		req.Total, req.TaxTotal, req.SubTotal,
	)
	if err != nil {
		return nil, fmt.Errorf("create_invoice insert: %w", err)
	}

	for i, item := range req.LineItems {
		itemID, _ := gonanoid.New()
		_, err = tx.Exec(`
			INSERT INTO invoiceLineItems (id, invoiceId, description, quantity, unitPrice, taxRate, position)
			VALUES (?, ?, ?, ?, ?, ?, ?)`,
			itemID, req.ID, item.Description, item.Quantity, item.UnitPrice, item.TaxRate, i,
		)
		if err != nil {
			return nil, fmt.Errorf("create_invoice line_item: %w", err)
		}
	}

	_, err = tx.Exec(
		`UPDATE organizations SET invoice_number_counter = invoice_number_counter + 1 WHERE id = ?`,
		req.OrganizationID,
	)
	if err != nil {
		return nil, fmt.Errorf("create_invoice counter: %w", err)
	}

	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("create_invoice commit: %w", err)
	}

	return d.GetInvoice(req.ID)
}

func (d *Database) UpdateInvoice(invoiceID string, updates UpdateInvoiceRequest) (*Invoice, error) {
	tx, err := d.DB.Beginx()
	if err != nil {
		return nil, fmt.Errorf("update_invoice begin: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	// Required fields use COALESCE (kept when nil); nullable fields are set
	// directly so users can clear them by passing null.
	_, err = tx.Exec(`
		UPDATE invoices
		SET number        = COALESCE(?, number),
		    state         = COALESCE(?, state),
		    clientId      = COALESCE(?, clientId),
		    date          = COALESCE(?, date),
		    dueDate       = ?,
		    currency      = COALESCE(?, currency),
		    customerNotes = ?,
		    overdueCharge = ?,
		    total         = COALESCE(?, total),
		    taxTotal      = COALESCE(?, taxTotal),
		    subTotal      = COALESCE(?, subTotal)
		WHERE id = ?`,
		updates.Number, updates.State, updates.ClientID,
		updates.Date, updates.DueDate, updates.Currency,
		updates.CustomerNotes, updates.OverdueCharge,
		updates.Total, updates.TaxTotal, updates.SubTotal,
		invoiceID,
	)
	if err != nil {
		return nil, fmt.Errorf("update_invoice exec: %w", err)
	}

	if updates.LineItems != nil {
		if _, err = tx.Exec(`DELETE FROM invoiceLineItems WHERE invoiceId = ?`, invoiceID); err != nil {
			return nil, fmt.Errorf("update_invoice delete_items: %w", err)
		}
		for i, item := range *updates.LineItems {
			itemID, _ := gonanoid.New()
			_, err = tx.Exec(`
				INSERT INTO invoiceLineItems (id, invoiceId, description, quantity, unitPrice, taxRate, position)
				VALUES (?, ?, ?, ?, ?, ?, ?)`,
				itemID, invoiceID, item.Description, item.Quantity, item.UnitPrice, item.TaxRate, i,
			)
			if err != nil {
				return nil, fmt.Errorf("update_invoice line_item: %w", err)
			}
		}
	}

	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("update_invoice commit: %w", err)
	}

	return d.GetInvoice(invoiceID)
}

func (d *Database) UpdateInvoiceState(invoiceID string, state string) (*Invoice, error) {
	_, err := d.DB.Exec(`UPDATE invoices SET state = ? WHERE id = ?`, state, invoiceID)
	if err != nil {
		return nil, fmt.Errorf("update_invoice_state: %w", err)
	}
	return d.GetInvoice(invoiceID)
}

func (d *Database) DeleteInvoice(invoiceID string) (bool, error) {
	tx, err := d.DB.Beginx()
	if err != nil {
		return false, fmt.Errorf("delete_invoice begin: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	if _, err = tx.Exec(`DELETE FROM invoiceLineItems WHERE invoiceId = ?`, invoiceID); err != nil {
		return false, fmt.Errorf("delete_invoice items: %w", err)
	}

	res, err := tx.Exec(`DELETE FROM invoices WHERE id = ?`, invoiceID)
	if err != nil {
		return false, fmt.Errorf("delete_invoice: %w", err)
	}

	if err := tx.Commit(); err != nil {
		return false, fmt.Errorf("delete_invoice commit: %w", err)
	}

	n, _ := res.RowsAffected()
	return n > 0, nil
}
